### PR TITLE
Other components support set log level dynamically

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What this PR does / why we need it**:

#63777 introduced a way to set glog.logging.verbosity dynamically. 
We should enable this for all other components, which is specially useful in debugging. 

**Release note**:

```release-note
Expose `/debug/flags/v` to allow kubelet dynamically set glog logging level.  If want to change glog level to 3, you only have to send a PUT request like `curl -X PUT http://127.0.0.1:8080/debug/flags/v -d "3"`.
```

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
cc @lavalamp @deads2k @sttts @CaoShuFeng ptal

File: cmd/kube-scheduler/app/server.go [hunk 1]
```diff
@@ -237,15 +237,9 @@ func installMetricHandler(pathRecorderMux *mux.PathRecorderMux) {
 }
 
 // newMetricsHandler builds a metrics server from the config.
-func newMetricsHandler(config *componentconfig.KubeSchedulerConfiguration) http.Handler {
+func newMetricsHandler() http.Handler {
 	pathRecorderMux := mux.NewPathRecorderMux("kube-scheduler")
 	installMetricHandler(pathRecorderMux)
-	if config.EnableProfiling {
```
Comment (k82cn) [inline cmd/kube-scheduler/app/server.go]: why remove this?

Comment (hzxuzhonghu) [reply cmd/kube-scheduler/app/server.go]: I think it's duplicated with healthz handler. Should not exist in both healthz and metrics handler.

Comment (k82cn) [reply cmd/kube-scheduler/app/server.go]: let's open a separate PR for that: it seems enabled by different params, did not get time to deep into the detail). 

The log part LGTM.

Comment (hzxuzhonghu) [reply cmd/kube-scheduler/app/server.go]: OK. will update right now.

File: cmd/kube-scheduler/app/server.go [hunk 2]
```diff
@@ -263,6 +274,16 @@ func newHealthzHandler(config *componentconfig.KubeSchedulerConfiguration, separ
 		if config.EnableContentionProfiling {
 			goruntime.SetBlockProfileRate(1)
 		}
+		// so far, only logging related endpoints are considered valid to add for these debug flags.
+		routes.DebugFlags{}.Install(pathRecorderMux, "v", routes.StringFlagPutHandler(
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply cmd/kube-scheduler/app/server.go]: SGTM

**Comment by thockin** [comment]
Shouldn't we have some sort of auth mechanism here?  This is a DoS waiting to happen.

**Comment by sttts** [comment]
> Shouldn't we have some sort of auth mechanism here? This is a DoS waiting to happen.

This has to be behind an RBAC rule by default only allowing it for the cluster admin, plus an e2e test to verify that. For the scheduler and controller-manager we will add delegated authn/z for 1.12 (WIP PRs are open). Not sure about kubelet and kube-proxy (/cc @liggitt).

**Comment by hzxuzhonghu** [comment]
> Not sure about kubelet and kube-proxy

kubelet now startup a secure https schema, and after authn/authz. Not sure about kube-proxy.

**Comment by liggitt** [comment]
> Not sure about kubelet and kube-proxy
> 
> kubelet now startup a secure https schema, and after authn/authz. Not sure about kube-proxy.
> 
> Just confirmed kube-proxy is insecure http without authn/authz.

can we enable this on the components that can protect it, and defer on the ones that can't yet?

**Comment by liggitt** [comment]
> @thockin @liggitt Now this only support kubelet, since others are without authn/authz

and apiserver, right?

**Comment by hzxuzhonghu** [comment]
>and apiserver, right?

Yes, it is already supported before. In this pr, as I move glog setter to a common place,  I just update apiserver to make use of the new package.

**Comment by hzxuzhonghu** [comment]
kindly ping @thockin @liggitt

**Comment by hzxuzhonghu** [comment]
Thanks @liggitt 

@thockin can you approve?

**Comment by hzxuzhonghu** [comment]
ping @lavalamp @thockin for approval

**Comment by sttts** [comment]
/approve

for apiextensions-apiserver.

**Comment by hzxuzhonghu** [comment]
ping @tallclair for kubelet part

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 3]
```diff
@@ -576,15 +576,7 @@ func installAPI(s *GenericAPIServer, c *Config) {
...
-		routes.DebugFlags{}.Install(s.Handler.NonGoRestfulMux, "v", routes.StringFlagPutHandler(
-			routes.StringFlagSetterFunc(func(val string) (string, error) {
-				var level glog.Level
-				if err := level.Set(val); err != nil {
-					return "", fmt.Errorf("failed set glog.logging.verbosity %s: %v", val, err)
-				}
-				return "successfully set glog.logging.verbosity to " + val, nil
-			}),
-		))
+		routes.DebugFlags{}.Install(s.Handler.NonGoRestfulMux, "v", routes.StringFlagPutHandler(logs.GlogSetter))
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: Why is this gated on EnableProfiling?

File: staging/src/k8s.io/apiserver/pkg/util/logs/logs.go [hunk 4]
```diff
@@ -67,3 +68,12 @@ func FlushLogs() {
...
 	return log.New(GlogWriter{}, prefix, 0)
 }
+
+// GlogSetter is a setter to set glog level.
+func GlogSetter(val string) (string, error) {
+	var level glog.Level
+	if err := level.Set(val); err != nil {
+		return "", fmt.Errorf("failed set glog.logging.verbosity %s: %v", val, err)
+	}
+	return "successfully set glog.logging.verbosity to " + val, nil
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/util/logs/logs.go]: nit: prefer
```
fmt.Sprintf("successfully set glog.logging.verbosity to %d", level)
```

Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/util/logs/logs.go]: Actually, `level.Set(val)` weirdly doesn't actually modify the level value, so nevermind...

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 3 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: xref: https://github.com/kubernetes/kubernetes/pull/63777#discussion_r191838475

File: staging/src/k8s.io/apiserver/pkg/util/logs/logs.go (hunk 4 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/util/logs/logs.go]: ok

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/util/logs/logs.go]: `fmt.Sprintf("successfully set glog.logging.verbosity to %s", val)`

**Comment by tossmilestone** [comment]
Does this feature can be used in kube-scheduler? @hzxuzhonghu

**Comment by hzxuzhonghu** [comment]
@tossmilestone ref: https://github.com/kubernetes/kubernetes/pull/64601#issuecomment-398589877

**Pull request merged**